File: main/setpoint.c
/*
 * setpoint.c — UDP Listener + Shared Setpoint State
 *
 * Binds a UDP socket on port 4210 (INADDR_ANY) and parses incoming
 * 12-byte packets into the shared setpoint_t struct.
 *
 * Task: Core 0, priority 4, 4 KB stack.
 * The IMU/flight tasks on Core 1 read via setpoint_get() — no contention.
 */

#include "setpoint.h"

#include <string.h>
#include <errno.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include "freertos/FreeRTOS.h"
#include "freertos/task.h"
#include "freertos/semphr.h"

#include "esp_log.h"
#include "esp_timer.h"

static const char *TAG = "[SETPOINT]";

// ============================================================
// Packet Layout Constants
// ============================================================
#define MAGIC_0         0xAB
#define MAGIC_1         0xCD
#define PACKET_SIZE     12

// ============================================================
// Task Configuration
// ============================================================
#define UDP_TASK_STACK  4096
#define UDP_TASK_PRIO   4       // Low — networking tier, Core 0
#define UDP_TASK_CORE   0

// ============================================================
// Internal State
// ============================================================
static SemaphoreHandle_t s_mutex      = NULL;
static int64_t           s_last_rx_us = 0;
static bool              s_started    = false;

static setpoint_t s_setpoint = {
    .throttle     = 1000,
    .roll         = 0.0f,
    .pitch        = 0.0f,
    .yaw_rate     = 0.0f,
    .flags        = 0,
    .fresh        = false,
    .packet_count = 0,
};

// ============================================================
// Packet Parser
// Returns true and fills *out on success; returns false and logs on error.
// ============================================================
static bool parse_packet(const uint8_t *buf, int len, setpoint_t *out)
{
    if (len != PACKET_SIZE) {
        ESP_LOGW(TAG, "Wrong length: got %d, expected %d", len, PACKET_SIZE);
        return false;
    }

    if (buf[0] != MAGIC_0 || buf[1] != MAGIC_1) {
        ESP_LOGW(TAG, "Bad magic: 0x%02X 0x%02X", buf[0], buf[1]);
        return false;
    }

    // XOR checksum over bytes [0..10], compare to byte [11]
    uint8_t chk = 0;
    for (int i = 0; i < PACKET_SIZE - 1; i++) chk ^= buf[i];
    if (chk != buf[PACKET_SIZE - 1]) {
        ESP_LOGW(TAG, "Checksum fail: calc=0x%02X recv=0x%02X", chk, buf[PACKET_SIZE - 1]);
        return false;
    }

    // Extract big-endian fields
    uint16_t thr = ((uint16_t)buf[2] << 8) | buf[3];
    int16_t  rol = (int16_t)(((uint16_t)buf[4] << 8) | buf[5]);
    int16_t  pit = (int16_t)(((uint16_t)buf[6] << 8) | buf[7]);
    int16_t  yaw = (int16_t)(((uint16_t)buf[8] << 8) | buf[9]);
    uint8_t  flg = buf[10];

    // Hard range validation — reject anything physically impossible
    if (thr < 1000 || thr > 2000) { ESP_LOGW(TAG, "Throttle OOB: %u", thr);  return false; }
    if (rol < -450  || rol > 450)  { ESP_LOGW(TAG, "Roll OOB: %d",   rol);   return false; }
    if (pit < -450  || pit > 450)  { ESP_LOGW(TAG, "Pitch OOB: %d",  pit);   return false; }
    if (yaw < -3000 || yaw > 3000) { ESP_LOGW(TAG, "Yaw OOB: %d",    yaw);   return false; }

    out->throttle = thr;
    out->roll     = rol / 10.0f;    // tenths-of-degree → degrees
    out->pitch    = pit / 10.0f;
    out->yaw_rate = yaw / 10.0f;
    out->flags    = flg;

    return true;
}

// ============================================================
// UDP Listener Task — Core 0
// ============================================================
static void udp_listener_task(void *arg)
{
    ESP_LOGI(TAG, "UDP listener starting on port %d (Core %d)",
             SETPOINT_UDP_PORT, UDP_TASK_CORE);

    // Create UDP socket
    int sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
    if (sock < 0) {
        ESP_LOGE(TAG, "socket() failed: errno %d", errno);
        vTaskDelete(NULL);
        return;
    }

    // 200 ms receive timeout — allows periodic staleness checks without spin-waiting
    struct timeval tv = { .tv_sec = 0, .tv_usec = 200000 };
    setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));

    // Address reuse — allows clean restart without TIME_WAIT issues
    int reuse = 1;
    setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof(reuse));

    struct sockaddr_in bind_addr = {
        .sin_family      = AF_INET,
        .sin_addr.s_addr = htonl(INADDR_ANY),
        .sin_port        = htons(SETPOINT_UDP_PORT),
    };

    if (bind(sock, (struct sockaddr *)&bind_addr, sizeof(bind_addr)) < 0) {
        ESP_LOGE(TAG, "bind() failed: errno %d — port %d may be in use",
                 errno, SETPOINT_UDP_PORT);
        close(sock);
        vTaskDelete(NULL);
        return;
    }

    ESP_LOGI(TAG, "Socket bound — waiting for packets on 0.0.0.0:%d", SETPOINT_UDP_PORT);

    // Receive buffer — slightly oversized to detect truncation
    uint8_t buf[PACKET_SIZE + 8];
    struct sockaddr_in sender;
    socklen_t sender_len = sizeof(sender);
    char sender_str[INET_ADDRSTRLEN];

    bool     first_packet   = true;
    uint32_t stale_log_tick = 0;     // Rate-limit stale log spam

    while (1) {
        int n = recvfrom(sock, buf, sizeof(buf), 0,
                         (struct sockaddr *)&sender, &sender_len);

        if (n < 0) {
            if (errno == EAGAIN || errno == EWOULDBLOCK) {
                // Normal 200 ms timeout — check for staleness
                int64_t now_us  = esp_timer_get_time();
                int64_t age_ms  = (s_last_rx_us > 0) ?
                                  (now_us - s_last_rx_us) / 1000LL : -1;

                if (age_ms > (int64_t)SETPOINT_TIMEOUT_MS) {
                    if (xSemaphoreTake(s_mutex, pdMS_TO_TICKS(5)) == pdTRUE) {
                        s_setpoint.fresh = false;
                        xSemaphoreGive(s_mutex);
                    }
                    // Log every ~2 seconds to avoid flooding serial
                    if (stale_log_tick % 10 == 0 && s_last_rx_us > 0) {
                        ESP_LOGW(TAG, "Setpoint STALE — %lld ms since last packet", age_ms);
                    }
                    stale_log_tick++;
                }
            } else {
                ESP_LOGE(TAG, "recvfrom error: errno %d", errno);
                vTaskDelay(pdMS_TO_TICKS(100));
            }
            continue;
        }

        // Successfully received a datagram — parse it
        setpoint_t parsed = {0};
        if (!parse_packet(buf, n, &parsed)) {
            continue;
        }

        // Log first valid packet's origin (once)
        if (first_packet) {
            inet_ntop(AF_INET, &sender.sin_addr, sender_str, sizeof(sender_str));
            ESP_LOGI(TAG, "First valid packet from %s:%d",
                     sender_str, ntohs(sender.sin_port));
            first_packet = false;
        }

        // Commit to shared state under mutex
        if (xSemaphoreTake(s_mutex, pdMS_TO_TICKS(5)) == pdTRUE) {
            s_setpoint.throttle     = parsed.throttle;
            s_setpoint.roll         = parsed.roll;
            s_setpoint.pitch        = parsed.pitch;
            s_setpoint.yaw_rate     = parsed.yaw_rate;
            s_setpoint.flags        = parsed.flags;
            s_setpoint.fresh        = true;
            s_setpoint.packet_count++;
            s_last_rx_us = esp_timer_get_time();
            xSemaphoreGive(s_mutex);
        }

        stale_log_tick = 0;

        // Periodic status log — every 100 packets (~2 s at 50 Hz)
        if (s_setpoint.packet_count % 100 == 0) {
            ESP_LOGI(TAG, "PKT #%lu | THR=%u µs | R=%.1f° P=%.1f° Y=%.1f°/s | flags=0x%02X",
                     (unsigned long)s_setpoint.packet_count,
                     s_setpoint.throttle,
                     s_setpoint.roll,
                     s_setpoint.pitch,
                     s_setpoint.yaw_rate,
                     s_setpoint.flags);
        }
    }

    // Never reached, but clean up if it ever is
    close(sock);
    vTaskDelete(NULL);
}

// ============================================================
// Public API
// ============================================================

esp_err_t setpoint_init(void)
{
    if (s_started) {
        ESP_LOGW(TAG, "Already initialised — ignoring second call");
        return ESP_OK;
    }

    s_mutex = xSemaphoreCreateMutex();
    if (!s_mutex) {
        ESP_LOGE(TAG, "Mutex creation failed — out of heap?");
        return ESP_ERR_NO_MEM;
    }

    BaseType_t rc = xTaskCreatePinnedToCore(
        udp_listener_task,
        "setpoint_udp",
        UDP_TASK_STACK,
        NULL,
        UDP_TASK_PRIO,
        NULL,
        UDP_TASK_CORE
    );

    if (rc != pdPASS) {
        ESP_LOGE(TAG, "Task creation failed");
        vSemaphoreDelete(s_mutex);
        s_mutex = NULL;
        return ESP_ERR_NO_MEM;
    }

    s_started = true;
    ESP_LOGI(TAG, "Setpoint subsystem ready — UDP port %d, Core %d, prio %d",
             SETPOINT_UDP_PORT, UDP_TASK_CORE, UDP_TASK_PRIO);
    return ESP_OK;
}

void setpoint_get(setpoint_t *out)
{
    if (!out) return;

    // If not yet initialised, return safe defaults without crashing
    if (!s_mutex) {
        out->throttle     = 1000;
        out->roll         = 0.0f;
        out->pitch        = 0.0f;
        out->yaw_rate     = 0.0f;
        out->flags        = 0;
        out->fresh        = false;
        out->packet_count = 0;
        return;
    }

    if (xSemaphoreTake(s_mutex, pdMS_TO_TICKS(5)) == pdTRUE) {
        memcpy(out, &s_setpoint, sizeof(setpoint_t));
        xSemaphoreGive(s_mutex);
    }
}

bool setpoint_is_fresh(void)
{
    if (!s_mutex) return false;
    bool fresh = false;
    if (xSemaphoreTake(s_mutex, pdMS_TO_TICKS(2)) == pdTRUE) {
        fresh = s_setpoint.fresh;
        xSemaphoreGive(s_mutex);
    }
    return fresh;
}

File: main/setpoint.py
#!/usr/bin/env python3
"""
test_setpoint.py — Phase 1 UDP setpoint verification tool
----------------------------------------------------------
Sends 12-byte UDP packets to the drone's ESP32-S3 on port 4210.
No external libraries required — uses Python stdlib only.

Usage:
    python3 test_setpoint.py <DRONE_IP>

Examples:
    python3 test_setpoint.py 192.168.1.105          # idle heartbeat
    python3 test_setpoint.py 192.168.1.105 --roll 15
    python3 test_setpoint.py 192.168.1.105 --thr 1200 --pitch -10 --yaw 30
    python3 test_setpoint.py 192.168.1.105 --hz 100  # stress test at 100 Hz

Packet format (12 bytes, big-endian):
    [0]    0xAB  magic
    [1]    0xCD  magic
    [2-3]  throttle  uint16  1000-2000 µs
    [4-5]  roll      int16   value*10  (±450 = ±45.0°)
    [6-7]  pitch     int16   value*10  (±450 = ±45.0°)
    [8-9]  yaw_rate  int16   value*10  (±3000 = ±300.0°/s)
    [10]   flags     uint8   reserved
    [11]   checksum  uint8   XOR of [0..10]
"""

import socket
import struct
import time
import sys
import argparse

DRONE_PORT = 4210
MAGIC_0    = 0xAB
MAGIC_1    = 0xCD
PACKET_SIZE = 12


def build_packet(throttle: int, roll_deg: float, pitch_deg: float,
                 yaw_rate_dps: float, flags: int = 0) -> bytes:
    """
    Build a validated 12-byte setpoint packet.
    Raises ValueError if any value is out of the accepted range.
    """
    throttle    = int(round(throttle))
    roll_raw    = int(round(roll_deg    * 10))
    pitch_raw   = int(round(pitch_deg   * 10))
    yaw_raw     = int(round(yaw_rate_dps * 10))

    if not (1000 <= throttle <= 2000):
        raise ValueError(f"Throttle {throttle} out of range [1000-2000]")
    if not (-450 <= roll_raw <= 450):
        raise ValueError(f"Roll {roll_deg}° out of range [±45°]")
    if not (-450 <= pitch_raw <= 450):
        raise ValueError(f"Pitch {pitch_deg}° out of range [±45°]")
    if not (-3000 <= yaw_raw <= 3000):
        raise ValueError(f"Yaw rate {yaw_rate_dps}°/s out of range [±300°/s]")

    # Pack header + payload (big-endian)
    # Format: BB H h h h B  →  magic0 magic1 throttle roll pitch yaw flags
    body = struct.pack('>BBHhhhB',
                       MAGIC_0, MAGIC_1,
                       throttle, roll_raw, pitch_raw, yaw_raw,
                       flags)
    assert len(body) == PACKET_SIZE - 1

    # XOR checksum over all 11 payload bytes
    checksum = 0
    for b in body:
        checksum ^= b

    return body + bytes([checksum])


def run(drone_ip: str, throttle: int, roll: float, pitch: float,
        yaw_rate: float, hz: float, duration_s: float):

    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    period = 1.0 / hz
    target = (drone_ip, DRONE_PORT)

    print(f"\n{'─'*54}")
    print(f"  Phase 1 setpoint test")
    print(f"  Target  : {drone_ip}:{DRONE_PORT}")
    print(f"  Rate    : {hz:.0f} Hz  (period {period*1000:.1f} ms)")
    print(f"  Throttle: {throttle} µs")
    print(f"  Roll    : {roll:.1f}°")
    print(f"  Pitch   : {pitch:.1f}°")
    print(f"  Yaw rate: {yaw_rate:.1f} °/s")
    if duration_s > 0:
        print(f"  Duration: {duration_s:.1f} s")
    else:
        print(f"  Duration: until Ctrl+C")
    print(f"{'─'*54}\n")

    pkt = build_packet(throttle, roll, pitch, yaw_rate)
    assert len(pkt) == PACKET_SIZE, f"Packet wrong size: {len(pkt)}"

    print(f"  Packet bytes: {pkt.hex().upper()}")
    print(f"  Checksum: 0x{pkt[-1]:02X}\n")
    print("  Sending... (Ctrl+C to stop)\n")

    sent = 0
    t_start = time.monotonic()
    next_send = t_start

    try:
        while True:
            now = time.monotonic()
            elapsed = now - t_start

            if duration_s > 0 and elapsed >= duration_s:
                break

            if now >= next_send:
                sock.sendto(pkt, target)
                sent += 1
                next_send += period

                # Progress update every 2 seconds
                if sent % max(1, int(hz * 2)) == 0:
                    rate = sent / elapsed if elapsed > 0.001 else 0
                    print(f"\r  Sent {sent:6d} packets  ({rate:.1f} Hz actual)    ",
                          end='', flush=True)

            # Sleep for ~half a period to avoid busy-wait, keep timing tight
            sleep_time = next_send - time.monotonic()
            if sleep_time > 0.001:
                time.sleep(sleep_time * 0.9)

    except KeyboardInterrupt:
        pass
    finally:
        elapsed = time.monotonic() - t_start
        sock.close()
        print(f"\n\n  ── Summary ──────────────────────────")
        print(f"  Sent   : {sent} packets")
        print(f"  Elapsed: {elapsed:.2f} s")
        print(f"  Rate   : {sent/elapsed:.1f} Hz actual")
        print(f"\n  Now check the ESP32 serial monitor for:")
        print(f"    [SETPOINT] First valid packet from {drone_ip}:...")
        print(f"    [SETPOINT] PKT #100 | THR={throttle} ...")
        print(f"\n  Stop sending and wait 0.5s to see:")
        print(f"    [SETPOINT] Setpoint STALE — ... ms since last packet")
        print(f"{'─'*38}\n")


def main():
    parser = argparse.ArgumentParser(
        description="Phase 1 UDP setpoint test — sends packets to ESP32 drone"
    )
    parser.add_argument("drone_ip", help="IP address of the drone's ESP32-S3")
    parser.add_argument("--thr",   type=int,   default=1000,
                        help="Throttle µs (1000-2000, default 1000)")
    parser.add_argument("--roll",  type=float, default=0.0,
                        help="Roll target degrees (±45, default 0)")
    parser.add_argument("--pitch", type=float, default=0.0,
                        help="Pitch target degrees (±45, default 0)")
    parser.add_argument("--yaw",   type=float, default=0.0,
                        help="Yaw rate °/s (±300, default 0)")
    parser.add_argument("--hz",    type=float, default=50.0,
                        help="Send rate Hz (default 50)")
    parser.add_argument("--time",  type=float, default=0.0,
                        help="Duration in seconds, 0 = run until Ctrl+C (default)")

    args = parser.parse_args()

    try:
        run(
            drone_ip   = args.drone_ip,
            throttle   = args.thr,
            roll       = args.roll,
            pitch      = args.pitch,
            yaw_rate   = args.yaw,
            hz         = args.hz,
            duration_s = args.time,
        )
    except ValueError as e:
        print(f"\n  ERROR: {e}\n")
        sys.exit(1)


if __name__ == "__main__":
    main()